Store/StoreBraintreePaymentProvider.php

```diff
@@ -0,0 +1,455 @@
+		}
+
+		$custom_fields = $this->getCustomFields($order);
+		if (count($custom_fields) > 0) {
+			$request['customFields'] = $custom_fields;
+		}
+
+		// do transaction
+		$this->setConfig();
+		$response = Braintree_Transaction::sale($request);
+
+		// check for errors and throw exception
+		if (!$response->success) {
+			throw new StorePaymentBraintreeException(
+				'Error processing Braintree payment.',
+				0,
+				$response
+			);
+		}
+
+		$class_name = SwatDBClassMap::get('StorePaymentMethodTransaction');
+		$transaction = new $class_name();
+
+		$transaction->transaction_type = StorePaymentRequest::TYPE_PAY;
+		$transaction->transaction_id = $response->transaction->id;
+		$transaction->createdate = new SwatDate();
+		$transaction->createdate->toUTC();
+
+		return $transaction;
+	}
+
+	// }}}
+	// {{{ public function getExceptionMessageId()
+
+	public function getExceptionMessageId(Exception $e)
+	{
+		if ($e instanceof Braintree_Exception_Authentication ||
+			$e instanceof Braintree_Exception_Authorization ||
+			$e instanceof Braintree_Exception_Configuration ||
+			$e instanceof Braintree_Exception_ServerError ||
+			$e instanceof Braintree_Exception_UpgradeRequired) {
+			return 'payment-error';
+		}
+
+		if ($e instanceof StorePaymentBraintreeException) {
+			$response = $e->getResponse();
+
+			// data validation error(s)
+			if ($response->errors->deepSize() > 0) {
+				return 'payment-error';
+			}
+
+			$transaction = $response->transaction;
+
+			// transaction error
+			if ($transaction->status === 'processor_declined') {
+				switch ($transaction->processorResponseCode) {
+				case 2004: // expired card
+					return 'card-expired';
+				case 2010: // issuer declined CVV
+					return 'card-verification-value';
+				case 2060: // address verification failed
+					return 'address-mismatch';
+				default:
+					return 'payment-error';
+				}
+			}
+
+			if ($transaction->status === 'settlement_declined') {
+				return 'payment-error';
+			}
+
+			if ($transaction->status === 'gateway_rejected') {
+				switch ($transaction->gatewayRejectionReason) {
+				case 'avs':
+					return 'address-mismatch';
+				case 'avs_and_cvv':
+				case 'cvv':
+					return 'card-verification-value';
+				default:
+					return 'payment-error';
+				}
+			}
+		}
+
+		return null;
+	}
+
+	// }}}
+	// {{{ protected function setConfig()
+
+	protected function setConfig()
+	{
+		Braintree_Configuration::environment($this->environment);
+		Braintree_Configuration::merchantId($this->merchant_id);
+		Braintree_Configuration::publicKey($this->public_key);
+		Braintree_Configuration::privateKey($this->private_key);
+	}
+
+	// }}}
+	// {{{ protected function getCreditCard()
+
+	/**
+	 * @sensitive $card_number
+	 * @sensitive $card_verification_value
+	 */
+	protected function getCreditCard(
+		StoreOrder $order, $card_number, $card_verification_value = null)
+	{
+		// get expiration date and cardholder from payment method
+		$date = new SwatDate('-1 month');
```

Comment:
Turns out this is a default date in case the payment method doesn't exist - this needs a note explaining what it is.